firmware/common/cpld_update.cpp

```diff
@@ -87,6 +92,121 @@ CpldUpdateStatus update_if_necessary(
     return ok ? CpldUpdateStatus::Success : CpldUpdateStatus::Program_failed;
 }
 
+static CpldUpdateStatus enter_maintenance_mode(CPLD& cpld) {
+    /* Unknown state */
+    cpld.reset();
+    cpld.run_test_idle();
+
+    /* Run-Test/Idle */
+    if (!cpld.idcode_ok()) {
+        return CpldUpdateStatus::Idcode_check_failed;
+    }
+
+    cpld.sample();
+    cpld.bypass();
+    cpld.enable();
+
+    /* If silicon ID doesn't match, there's a serious problem. Leave CPLD
+     * in passive state.
+     */
+    if (!cpld.silicon_id_ok()) {
+        return CpldUpdateStatus::Silicon_id_check_failed;
+    }
+
+    return CpldUpdateStatus::Success;
+}
+
+static void exit_maintenance_mode(CPLD& cpld) {
+    cpld.disable();
+    cpld.bypass();
+
+    /* Initiate SRAM reload from flash we just programmed. */
+    cpld.sample();
+    cpld.clamp();
+    cpld.disable();
+}
+
+static uint32_t get_firmware_crc(CPLD& cpld) {
+    CRC<32> crc{0x04c11db7, 0xffffffff, 0xffffffff};
+    cpld.prepare_read(0x0000);
+
+    for (size_t i = 0; i < 3328; i++) {
+        uint16_t data = cpld.read();
+        crc.process_byte((data >> 0) & 0xff);
+        crc.process_byte((data >> 8) & 0xff);
+        crc.process_byte((data >> 16) & 0xff);
```

Comment:
Minor comment:

Just slightly confusing here that cpld.read() returns a U32, then it's stuffed into "U16 data" here because apparently we know that only the low 16 bits are valid.  But if we are using a U16, then why do we bother with the (data >> 16) and (data >> 24), versus just calling crc.process_byte(0)?

And, since CRC32 is so much slower than simply adding up the bytes (i.e. in the "simple_checksum" function), it does seem like those additional calls to crc_process_byte(0) could be skipped entirely ((in which case you'd of course need to know what the new checksum values are)

Reply:
oh you are right. I didn't pay attention as I stole that code from the usb serial command -.- I'll leave it like that for now, since it is a bit troublsome to get the checksums and the runtime is not an issue for that few bytes.